Main/SongSelect.cpp

```diff
@@ -1077,42 +1106,63 @@ class SongSelect_Impl : public SongSelect
 			g_application->DisposeLua(m_lua);
 	}
 
-	// When a map is selected in the song wheel
-	void OnMapSelected(MapIndex* map)
+	void m_updatePreview(DifficultyIndex *diff, bool mapChanged)
 	{
-		if (map == m_currentPreviewAudio){
-			if (m_previewDelayTicks){
-				--m_previewDelayTicks;
-			}else if (!m_previewLoaded && !m_currentPreviewAudio->difficulties.empty()){
-				// Set current preview audio
-				DifficultyIndex* previewDiff = m_currentPreviewAudio->difficulties[0];
-				String audioPath = m_currentPreviewAudio->path + Path::sep + previewDiff->settings.audioNoFX;
+		String mapRootPath = diff->path.substr(0, diff->path.find_last_of(Path::sep));
 
-				AudioStream previewAudio = g_audio->CreateStream(audioPath);
-				if (previewAudio)
-				{
-					previewAudio->SetPosition(previewDiff->settings.previewOffset);
-					m_previewPlayer.FadeTo(previewAudio);
-				}
-				else
-				{
-					Logf("Failed to load preview audio from [%s]", Logger::Warning, audioPath);
-					m_previewPlayer.FadeTo(AudioStream());
-				}
-				m_previewLoaded = true;
-				// m_previewPlayer.Restore();
+		// Set current preview audio
+		String audioPath = diff->settings.previewFile.length() > 0 ?
+			mapRootPath + Path::sep + diff->settings.previewFile :
+			mapRootPath + Path::sep + diff->settings.audioNoFX;
+
+		AudioStream previewAudio = g_audio->CreateStream(audioPath);
```

Comment:
Shouldn't `newPreview` be checked before attempting to create the new preview audio, so it doesn't load files that it will not end up using?

Reply:
Yeah, good catch. I actually have that change made on my local but wasn't gonna commit until I solved the leak.

Reply:
I've committed a fix for this.